### PR TITLE
Improves spiffe-csi-driver security isolation in k8s deployment

### PR DESCRIPTION
**Pull request by alexandrealvino**
Updates the `spiffe-csi-driver` k8s deployment in order to improve its security isolation.

Changes:

- The spiffe-csi driver will run as a DaemonSet in a single pod instead of in the spire-agent pod.
- The spiffe-csi driver will have its own service account instead of using the same one used by the spire-agent. The spire-agent DaemonSet RBAC is restrictive only to get verb to the following resources: pods, nodes, and nodes/proxy. The spiffe-csi driver ’s service account will have default behavior, thus more restrictive.
- Configured pod securityContext to limit the scope of allowed capabilities of the spiffe-csi driver using the following spec:

```yaml
securityContext:
  readOnlyRootFilesystem: true
  capabilities:
    drop:
      - all
  privileged: true
```

Signed-off-by: Alexandre Alvino <alexandre.alvino@hpe.com>

### CODE REVIEW

**Comment by faisal-memon** [comment]
I think we should split the csi driver into its own file.

File: example/config/spiffe-csi-driver.yaml [hunk 1]
```diff
@@ -1,3 +1,117 @@
...
+    metadata:
+      namespace: spire
+      labels:
+        app: spiffe-csi-driver
+    spec:
+      serviceAccountName: spiffe-csi-driver
+      containers:
+        # This is the container which runs the SPIFFE CSI driver.
+        - name: spiffe-csi-driver
+          image: ghcr.io/spiffe/spiffe-csi-driver:nightly
```
Comment (faisal-memon) [inline example/config/spiffe-csi-driver.yaml:34]:
```suggestion
          image: ghcr.io/spiffe/spiffe-csi-driver:0.2.0
```

File: example/config/spiffe-csi-driver.yaml [hunk 2]
```diff
@@ -1,3 +1,117 @@
...
+            readOnlyRootFilesystem: true
+            capabilities:
+              drop:
+                - all
+            privileged: true
+        # This container runs the CSI Node Driver Registrar which takes care
+        # of all the little details required to register a CSI driver with
+        # the kubelet.
+        - name: node-driver-registrar
+          image: quay.io/k8scsi/csi-node-driver-registrar:v2.0.1
```
Comment (faisal-memon) [inline example/config/spiffe-csi-driver.yaml:71]: This is up to 2.6.0. Should we bump to the latest version?

File: example/config/spiffe-csi-driver.yaml [hunk 3]
```diff
@@ -1,3 +1,117 @@
...
+            # The registrar needs access to the SPIFFE CSI driver socket
+            - mountPath: /spiffe-csi
+              name: spiffe-csi-socket-dir
+            # The registrar needs access to the Kubelet plugin registration
+            # directory
+            - name: kubelet-plugin-registration-dir
+              mountPath: /registration
+      volumes:
+        # This volume is used to share the Workload API socket between the CSI
+        # driver and SPIRE agent. Note, an emptyDir volume could also be used,
```
Comment (azdagron) [inline example/config/spiffe-csi-driver.yaml:87]: With this split out, we need a small clarification added to this comment:

```
... an emptyDir volume could also be used (if the CSI driver and SPIRE agent shared a pod), however...`

**Comment by azdagron** [comment]
This is a good improvement. I like it. Thanks, @alexandrealvino !

**Comment by azdagron** [comment]
Can we make a similar set of changes to the test/config/ directory (and update kustomization.yaml)? We want to make sure these changes are tested via CI/CD.

File: example/config/spiffe-csi-driver.yaml (hunk 2 above)
Comment (alexandrealvino) [reply example/config/spiffe-csi-driver.yaml:71]: I believe we can hop to the newest version, I will validate it

**Comment by alexandrealvino** [comment]
Thanks for the comments @faisal-memon @azdagron. I've submited the changes, you can check it out when you guys have some spare time

File: example/config/spiffe-csi-driver.yaml [hunk 4]
```diff
@@ -1,3 +1,117 @@
+---
+
```
Comment (faisal-memon) [inline example/config/spiffe-csi-driver.yaml:2]: Do we need the leading `---`?  The other files don't have it.

File: example/config/spiffe-csi-driver.yaml [hunk 5]
```diff
@@ -1,3 +1,117 @@
...
+kind: DaemonSet
+metadata:
+  name: spiffe-csi-driver
+  namespace: spire
+  labels:
+    app: spiffe-csi-driver
+spec:
+  selector:
+    matchLabels:
+      app: spiffe-csi-driver
```
Comment (faisal-memon) [inline example/config/spiffe-csi-driver.yaml:23]: Is this needed if we don't have a matching `Service`?

File: example/config/spiffe-csi-driver.yaml (hunk 4 above)
Comment (alexandrealvino) [reply example/config/spiffe-csi-driver.yaml:2]: That's not necessary

File: example/config/spiffe-csi-driver.yaml (hunk 5 above)
Comment (alexandrealvino) [reply example/config/spiffe-csi-driver.yaml:23]: This spec is required to deploy the `DaemonSet`

File: example/config/spiffe-csi-driver.yaml [hunk 6]
```diff
@@ -1,3 +1,115 @@
...
+  template:
+    metadata:
+      namespace: spire
+      labels:
+        app: spiffe-csi-driver
+    spec:
+      serviceAccountName: spiffe-csi-driver
+      containers:
+        - name: spiffe-csi-driver
```
Comment (faisal-memon) [inline example/config/spiffe-csi-driver.yaml:30]:
```suggestion
        # This is the container which runs the SPIFFE CSI driver.
        - name: spiffe-csi-driver
```

Comment (faisal-memon) [reply example/config/spiffe-csi-driver.yaml:30]: Adding the comment back in. Forgot there are two containers in this deployment

File: test/config/spiffe-csi-driver.yaml [hunk 7]
```diff
@@ -1,22 +1,132 @@
...
+      app: spiffe-csi-driver
+  template:
+    metadata:
+      namespace: spire-system
+      labels:
+        app: spiffe-csi-driver
+    spec:
+      serviceAccountName: spiffe-csi-driver
+      containers:
+        - name: spiffe-csi-driver
```
Comment (faisal-memon) [inline test/config/spiffe-csi-driver.yaml:30]:
```suggestion
        # This is the container which runs the SPIFFE CSI driver.
        - name: spiffe-csi-driver
```

File: example/config/spiffe-csi-driver.yaml [hunk 8]
```diff
@@ -1,3 +1,116 @@
...
+          # This is the container which runs the SPIFFE CSI driver.
```
Comment (faisal-memon) [inline example/config/spiffe-csi-driver.yaml:31]: nit: comment should be above `-name:`

**Comment by faisal-memon** [comment]
Thanks for making the changes @alexandrealvino. It looks good to me.

**Comment by azdagron** [comment]
With this separated into two DaemonSets, how is this impacted by lifecycle? How does the spiffe-csi-driver behave if the agent DaemonSet is not running or takes a while to come up? Have we performed any testing around this?

**Comment by alexandrealvino** [comment]
> With this separated into two DaemonSets, how is this impacted by lifecycle? How does the spiffe-csi-driver behave if the agent DaemonSet is not running or takes a while to come up? Have we performed any testing around this?

There is no impact, it has the same behavior as expected. I've tested in multiple orders of installation, then restarting both pods in different orders as well, no errors occurred on the testing. Let me know if there is a particular test missing or evidence.

We already have this spec installed on a dev cluster.

**Review by azdagron — APPROVED** [review]
Thanks!

**Pull request merged**